pkg/controller/machineset/machineset_controller_test.go

```diff
@@ -0,0 +1,447 @@
+			ExpectMatch:   false,
+			ExpectMessage: "Expected no match: available host has non-matching label value",
+		},
+		{
+			Host: &bmh.BareMetalHost{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "host1",
+					Namespace: "default",
+					Labels:    map[string]string{"size": "large"},
+				},
+				Spec: bmh.BareMetalHostSpec{
+					ConsumerRef: &v1.ObjectReference{
+						Kind:       "Machine",
+						APIVersion: machinev1beta1.SchemeGroupVersion.String(),
+						Name:       "machine1",
+						Namespace:  "default",
+					},
+				},
+			},
+			Machines: []runtime.Object{
+				&machinev1beta1.Machine{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "machine1",
+						Namespace: "default",
+						Labels:    map[string]string{"machine.openshift.io/cluster-api-machineset": "cluster0-storage"},
+					},
+				},
+			},
+			HSelector: labels.SelectorFromSet(map[string]string{
+				"size": "extralarge",
+			}),
+			MSSelector: labels.SelectorFromSet(map[string]string{
+				"machine.openshift.io/cluster-api-machineset": "cluster0-storage",
+			}),
+			ExpectMatch:   true,
+			ExpectMessage: "Expected match: host consumer is a Machine that matches the MSSelector, even though the host has a non-matching label value",
+		},
+		{
+			Host: &bmh.BareMetalHost{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "host1",
+					Namespace: "default",
+					Labels:    map[string]string{"size": "large"},
+				},
+				Spec: bmh.BareMetalHostSpec{
+					ConsumerRef: &v1.ObjectReference{
+						Kind:       "Machine",
+						APIVersion: machinev1beta1.SchemeGroupVersion.String(),
+						Name:       "machine1",
+						Namespace:  "default",
+					},
+				},
+			},
+			Machines: []runtime.Object{
+				&machinev1beta1.Machine{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "machine1",
+						Namespace: "default",
+						Labels:    map[string]string{"machine.openshift.io/cluster-api-machineset": "cluster0-workers"},
+					},
+				},
+			},
+			HSelector: labels.SelectorFromSet(map[string]string{
+				"size": "large",
+			}),
+			MSSelector: labels.SelectorFromSet(map[string]string{
+				"machine.openshift.io/cluster-api-machineset": "cluster0-storage",
+			}),
+			ExpectMatch:   false,
+			ExpectMessage: "Expected no match: host consumer is a Machine that does not match the MSSelector",
+		},
+		{
+			Host: &bmh.BareMetalHost{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "host1",
+					Namespace: "default",
+					Labels:    map[string]string{"size": "large"},
+				},
+				Spec: bmh.BareMetalHostSpec{
+					ConsumerRef: &v1.ObjectReference{
+						Kind:       "NotAMachine",
+						APIVersion: machinev1beta1.SchemeGroupVersion.String(),
+						Name:       "machine1",
+						Namespace:  "default",
+					},
+				},
+			},
+			HSelector: labels.SelectorFromSet(map[string]string{
+				"size": "large",
+			}),
+			MSSelector:    labels.NewSelector(),
+			ExpectMatch:   false,
+			ExpectMessage: "Expected no match: host consumer is not a Machine",
```

Comment:
Another test case worth adding is when Kind == "Machine" but consumer.APIVersion != machinev1beta1.SchemeGroupVersion.String().